staging/src/k8s.io/client-go/tools/cache/controller.go

```diff
@@ -585,6 +605,91 @@ func processDeltas(
 	return nil
 }
 
+// processDeltasInBatch applies a batch of Deltas to the given Store and
+// notifies the ResourceEventHandler of add, update, or delete events.
+//
+// If the Store supports transactions (TransactionStore), all Deltas are applied
+// atomically in a single transaction and corresponding handler callbacks are
+// executed afterward. Otherwise, each Delta is processed individually.
+//
+// Returns an error if any Delta or transaction fails. For TransactionError,
+// only successful operations trigger callbacks.
+func processDeltasInBatch(
+	handler ResourceEventHandler,
+	clientState Store,
+	deltasList []Deltas,
+	isInInitialList bool,
+) error {
+	// from oldest to newest
+	txns := make([]Transaction, 0)
+	callbacks := make([]func(), 0)
+	txnStore, txnSupported := clientState.(TransactionStore)
+	if !txnSupported {
+		var errs []error
+		for _, delta := range deltasList {
+			if err := processDeltas(handler, clientState, delta, isInInitialList); err != nil {
+				errs = append(errs, err)
+			}
+		}
+		if len(errs) > 0 {
+			return fmt.Errorf("unexpected error when handling deltas: %v", errs)
+		}
+		return nil
+	}
+	// deltasList is a list of unique objects
+	for _, deltas := range deltasList {
+		for _, d := range deltas {
+			obj := d.Object
+			switch d.Type {
+			case Sync, Replaced, Added, Updated:
+				// it will only return one old object for each because items are unique
+				if old, exists, err := clientState.Get(obj); err == nil && exists {
+					txn := Transaction{
+						Type:   TransactionTypeUpdate,
+						Object: obj,
+					}
+					txns = append(txns, txn)
+					callbacks = append(callbacks, func() {
+						handler.OnUpdate(old, obj)
+					})
+				} else {
+					txn := Transaction{
+						Type:   TransactionTypeAdd,
+						Object: obj,
+					}
+					txns = append(txns, txn)
+					callbacks = append(callbacks, func() {
+						handler.OnAdd(obj, isInInitialList)
+					})
+				}
+			case Deleted:
+				txn := Transaction{
+					Type:   TransactionTypeDelete,
+					Object: obj,
+				}
+				txns = append(txns, txn)
+				callbacks = append(callbacks, func() {
+					handler.OnDelete(obj)
+				})
+			}
+		}
+	}
+
+	err := txnStore.Transaction(txns...)
+	var txnErr TransactionError
+	if errors.As(err, &txnErr) {
+		// if txn had error, only execute the callbacks for the successful ones
+		for _, SuccessfulIdx := range txnErr.SuccessfulIndices {
+			callbacks[SuccessfulIdx]()
+		}
+		return err
+	}
+	for _, callback := range callbacks {
+		callback()
+	}
+	return err
+}
```

Comment:
right now `func (c *cache) Transaction(txns ...Transaction) error {` only returns a no `TransactionError` if the cacheStorage is not ThreadSafeStoreWithTransaction.

The code right now will execute all callbacks and return the error independently, should we make this clear and handle also the error to avoid depending on current implementation and define better the contract?